### PR TITLE
Museum pipeline

### PR DESCRIPTION
**Pull request by TaliVeith**
This is a new sample-type, medmuseum. 

If specified, there are two new steps (both dead ends): 
1) 0055-rrna: maps unmerged reads against ribosomal RNA and gives out a % of rRNA mapped reads. 
2) 0075-mrna: maps merged reads against a selection of mRNAs

### CODE REVIEW

**Comment by TaliVeith** [comment]
Hi Terry, I put in the discussed changes. Can I go ahead? :)

File: 0055-rrna/rrna.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        depthList = []
+        for line in cd:
+            for index in range(region[0], region[1]):
+                if str(index) == line.split(':')[0]:
+                    depthList.append(int(line.split()[1]))
+        if len(depthList) == 0:
+            averageCoverage = 0
+        else:
+            averageCoverage = sum(depthList) / len(depthList)
+        return averageCoverage
```
Comment (terrycojones) [inline 0055-rrna/rrna.py]: Your code is fine, but it makes the reader do more thinking than is needed. In general, avoid putting code into a `with` section that doesn't need to be there (just as you do the same with a `try` section). I'd write it like this:

```suggestion
    depthList = []
    with open(coverageDepthFile) as cd:
        for line in cd:
            for index in range(region[0], region[1]):
                if str(index) == line.split(':')[0]:
                    depthList.append(int(line.split()[1]))
                    break  # Right?

    return sum(depthList) / len(depthList) if depthList else 0
```

But it looks like it could be even simpler:

```suggestion
    high, low = map(int, region[:2])
    depthList = []
    with open(coverageDepthFile) as cd:
        for line in cd:
            if low <= int(line.split(':', 1)[0]) <= high:
                depthList.append(int(line.split()[1]))
    
    return sum(depthList) / len(depthList) if depthList else 0
```

I find it hard to follow the logic of your code because I don't know what's in `region` (is it length 2, does it have ints or floats?) and I don't know what's in `line.split(':', 1)[0]`  (note that I added a `1` to the split, to have splitting stop after just one delimiter is found), and I don't know what `line.split()[1]` is supposed to contain (is it an `int`?).  So I'm not very confident in rewriting it or in reading it - because I don't know what's going on in the data I can't check its logic in any way.

File: 0055-rrna/rrna.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+               for rDNA in sDna45]
+    averageRdna = sum(allRdna) / len(allRdna) if allRdna else 0.0
+
+    if averageRrna == 0:
+        print('No reads mapped against rRNA! Exiting...', file=sys.stderr)
+        exit()
+
+    correctionCoeff = averageRdna / averageRrna
+
+    print('Percent of mapped reads, corrected for rDNA mapping:', file=fp)
```
Comment (terrycojones) [inline 0055-rrna/rrna.py:117]: The `with open(outfile, 'w') as fp:` line could come right before these two lines. The way you have the code atm, you force the reader (who will be you at some future point) to keep in mind that we're in the context of a file opened for writing 18 lines of code ago. So as I read the code I keep wondering when you're going to use `fp`, and finally you do. Better to put all this together. These tiny things really do add some cognitive overload to reading and reasoning about the code.  It matters little in isolated instances but it's good practice to keep the logic tighter (i.e., not so separated on the page, in the code) and easier to keep track of for more complicated situations.

File: 0055-rrna/rrna.sh [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+{
+    # Copy our input FASTQ to our output unchanged.
+    cp $fastq $out
+}
+
+# Could delete the loops - there is only one database here!
+function rrna()
+{
+	local sam=$task.sam
+	local bam=$task.bam
```
Comment (terrycojones) [inline 0055-rrna/rrna.sh]: Two extra spaces of indent here for some reason?

File: 0055-rrna/rrna.sh [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        echo "  bwa mem (against $bwaDatabaseName) started at $(date)" >> $log
+        bwa mem -t $nproc $bwaDatabase $fastq > $sam
+        echo "  bwa mem (against $bwaDatabaseName) stopped at $(date)" >> $log
+
+        samtools quickcheck $sam
+
+        # Convert sam file to bam file, sort, index.
+        samtools view -S -b $sam > $bam
+        samtools sort $bam > $sortedbam
+        samtools index $sortedbam
```
Comment (terrycojones) [inline 0055-rrna/rrna.sh:79]: Remove `$bam`?

File: 0055-rrna/rrna.sh [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+        # Extract mapped and unmapped from samfile.
+        samtools fastq -f 4 $sam > $task.unmapped
+        samtools fastq -F 4 $sam > $task.mapped
+
+        # Try sam-coverage-depth.py
```
Comment (terrycojones) [inline 0055-rrna/rrna.sh]: This is a useless comment :-)  Delete.

File: 0055-rrna/sbatch.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+set -Eeuo pipefail
+
+. ../common.sh
+
+task=$1
+log=$logDir/sbatch.log
+
+# NOTE!! The following must have the identical value set and used in rrna.sh
+out=$task.done
```
Comment (terrycojones) [inline 0055-rrna/sbatch.sh]: This still differs from the value in `rrna.sh`, doesn't it?

File: 0075-mrna/mrna.sh [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+function skip()
+{
+    # Copy our input FASTQ to our output unchanged.
+    cp $fastq $out
+}
+
+# Could delete the loops - there is only one database here!
+function mrna()
+{
+	local sam=$task.sam
```
Comment (terrycojones) [inline 0075-mrna/mrna.sh]: Indent.

File: 0075-mrna/sbatch.sh [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+set -Eeuo pipefail
+
+. ../common.sh
+
+task=$1
+log=$logDir/sbatch.log
+
+# NOTE!! The following must have the identical value set and used in mrna.sh
+out=$task.done
```
Comment (terrycojones) [inline 0075-mrna/sbatch.sh]: Ditto.

File: specification.json [hunk 9]
```diff
@@ -11,6 +11,12 @@
             "cwd": "005-trim",
             "script": "./sbatch.sh"
         },
+        {
+            "name": "rrna",
```
Comment (terrycojones) [inline specification.json:15]: I think `rrna` and `mrna` should be added as dependencies to the `stop` step.  Otherwise there is a chance it will run before they are done. This could happen if all other steps didn't need to be run (because they've already been run in the past). In which case the `stop` step will not wait for `rrna` or `mrna` and will create the `slurm-pipeline.done` file before the pipeline has finished.

File: 0055-rrna/rrna.py (hunk 2 above)
Comment (TaliVeith) [reply 0055-rrna/rrna.py:117]: yes, makes sense. Done

File: 0055-rrna/rrna.sh (hunk 3 above)
Comment (TaliVeith) [reply 0055-rrna/rrna.sh]: hm, this was a tab including 4 spaces, so should have been the same (and appears same in sublime text). Changed all indentation to spaces.

File: 0055-rrna/rrna.sh (hunk 4 above)
Comment (TaliVeith) [reply 0055-rrna/rrna.sh:79]: done

File: 0055-rrna/rrna.sh (hunk 5 above)
Comment (TaliVeith) [reply 0055-rrna/rrna.sh]: done

File: 0055-rrna/sbatch.sh (hunk 6 above)
Comment (TaliVeith) [reply 0055-rrna/sbatch.sh]: Ah yes. I got confused and thought it wasn't necessary after all. Got it now, done. I also got rid of $outUncompressed in rrna.sh, as there is never a need to compress the outfile.

File: 0075-mrna/mrna.sh (hunk 7 above)
Comment (TaliVeith) [reply 0075-mrna/mrna.sh]: done

File: 0075-mrna/sbatch.sh (hunk 8 above)
Comment (TaliVeith) [reply 0075-mrna/sbatch.sh]: Done. Also adjusted the use of $out and $outUncompressed in mrna.sh

File: specification.json (hunk 9 above)
Comment (TaliVeith) [reply specification.json:15]: Done
